### DISCUSSION TITLE
Re-declaration of Params

### DISCUSSION THREAD

**Discussion by steam-simulator** [idaholab/moose#23168] [Q&A General] [answered]
Hi, I have a trivial question.

When inheriting from a parent class, do I have to re-declare parameters common to the parent and child class in the child class? I'm getting the following declaration error, but I don't want to duplicate these parameters in the child class.  

error: '_M' was not declared in this scope

**Comment by GiudGiud** [comment]
Hello

no you should have not have to, you can ininitalize the input parameters using the validParams() from the parent class.

however here you seem to be asking about a class attribute. You should not have to either


Guillaume

**Comment by hugary1995** [comment, marked as answer]
```
error: '_M' was not declared in this scope
```
The error message is precise: the compiler cannot find a member named `_M`. Two typical causes of that are:
1. You forgot to declare that member in the header.
2. `_M` is supposed to be inherited from the parent class, but `_M` is private in the parent class.

**Reply by steam-simulator** [reply]
Hi @GiudGiud @hugary1995:

I can’t seem to figure a way out of this issue. `_M` is protected in the parent class. Since I’m inheriting, I should not declare it in the child. I’m still getting the declaration error. Here is the part of my header showing the inheritance:
```
template <bool is_ad>
class UpdatedMflowTempl : public MflowBaseTempl<is_ad>
```

**Reply by steam-simulator** [reply]
ok. i got it now

I should have use the following in the header: `using MflowBaseTempl<is_ad>::_M;`

**Reply by hugary1995** [reply]
Ha, yes. That's unfortunately necessary as `_M` is a non-dependent name, i.e., it doesn't depend on any template argument. The compiler tries to look it up early before template instantiation. To make it dependent you'll have to do one of the following:
1. `using MflowBaseTempl<is_ad>::_M;` as you've discovered.
2. use `this->_M`
3. use `MflowBaseTempl<is_ad>::_M`
